src/hookproxy/proxydll.cpp

```diff
@@ -0,0 +1,251 @@
+ *
+ * @copyright Thyme is free software: you can redistribute it and/or
+ *            modify it under the terms of the GNU General Public License
+ *            as published by the Free Software Foundation, either version
+ *            2 of the License, or (at your option) any later version.
+ *            A full copy of the GNU General Public License can be found in
+ *            LICENSE
+ */
+#include "hooker.h"
+#include "sha1.h"
+#include <cstdio>
+#include <cstdlib>
+#include <io.h>
+#include <fstream>
+#include <windows.h>
+
+/**
+ * Pointer for the hooking function from the payload.
+ */
+void (*Setup_Hooks)();
+
+/**
+ * Functions we will export that will jump to the original dll functions that we are proxying for injection.
+ * avifil32.dll in this case.
+ */
+FARPROC OriginalAVIStreamWrite;
+FARPROC OriginalAVIFileInit;
+FARPROC OriginalAVIFileOpenA;
+FARPROC OriginalAVIFileRelease;
+FARPROC OriginalAVIFileExit;
+FARPROC OriginalAVIFileCreateStreamA;
+FARPROC OriginalAVIStreamSetFormat;
+FARPROC OriginalAVIStreamRelease;
+__declspec(naked) void FakeAVIStreamWrite()
+{
+    _asm { jmp[OriginalAVIStreamWrite] }
+}
+__declspec(naked) void FakeAVIFileInit()
+{
+    _asm { jmp[OriginalAVIFileInit] }
+}
+__declspec(naked) void FakeAVIFileOpenA()
+{
+    _asm { jmp[OriginalAVIFileOpenA] }
+}
+__declspec(naked) void FakeAVIFileRelease()
+{
+    _asm { jmp[OriginalAVIFileRelease] }
+}
+__declspec(naked) void FakeAVIFileExit()
+{
+    _asm { jmp[OriginalAVIFileExit] }
+}
+__declspec(naked) void FakeAVIFileCreateStreamA()
+{
+    _asm { jmp[OriginalAVIFileCreateStreamA] }
+}
+__declspec(naked) void FakeAVIStreamSetFormat()
+{
+    _asm { jmp[OriginalAVIStreamSetFormat] }
+}
+__declspec(naked) void FakeAVIStreamRelease()
+{
+    _asm { jmp[OriginalAVIStreamRelease] }
+}
+
+/**
+ * Get the path for the exe we were loaded by.
+ */
+wchar_t *Get_Path()
+{
+    wchar_t buffer[MAX_PATH];
+    wchar_t *path = nullptr;
+    wchar_t *full_path = nullptr;
+
+    while (true) {
+        DWORD size;
+
+        size = GetModuleFileNameW(nullptr, buffer, sizeof(buffer) / sizeof(buffer[0]));
+
+        if (size == 0) {
+            break;
+        } else if (size == (DWORD)(sizeof(buffer) / sizeof(buffer[0]))) {
+            DWORD size_ = size;
+            do {
+                wchar_t *path_ = (wchar_t *)std::realloc(path, sizeof(wchar_t) * size_ * 2);
+
+                if (path_ == nullptr) {
+                    break;
+                }
+
+                size_ *= 2;
+                path = path_;
+                size = GetModuleFileNameW(nullptr, path, size_);
+            } while (size == size_);
+
+            if (size == size_) {
+                break;
+            }
+        } else {
+            path = buffer;
+        }
+
+        full_path = _wfullpath(nullptr, path, MAX_PATH);
+
+        break;
+    }
+
+    if (path != buffer) {
+        free(path);
+    }
+
+    return full_path;
```

Comment:
Is there a chance function can return non-null but `path` was freed above? If so, then user may double free?